### PR TITLE
Updated Keras and TensorFlow Task Runner and related workspaces.

### PR DESCRIPTION
**Pull request by tanwarsh**
Related issue https://github.com/securefederatedai/openfl/issues/973.

Changes in this PR:
1. Updated Tensorflow and Keras to Tensorflow v2.18.0 and Keras 3
2. Updated openfl/federated/task/runner_kr.py
3. Standardizing the workspace.
4. Removed workspaces.
6. Updated workspaces.
    1. keras_cnn_mnist
    2. keras_nlp 
    3. keras_2dunet

### CODE REVIEW

**Comment by tanwarsh** [comment]
I will add code changes for the remaining Keras workspace - Keras_cnn_with_compression and keras_nlp.

@teoparvanov @kta-intel @psfoley, I was facing an issue with optimizer weights due to resetting the weights before saving them to the tensor database and using them to rebuild the model, which resulted in no improvement in accuracy. I have skipped the rebuild (and hence resetting the model optimizer weights) for the first round. This can be further discussed, and we can proceed with these changes once all the pipelines are working and the workspace changes are also completed.

File: .github/workflows/ubuntu.yml [hunk 1]
```diff
@@ -32,10 +32,10 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v3
-    - name: Set up Python 3.8
+    - name: Set up Python 3.9
```
Comment (teoparvanov) [inline .github/workflows/ubuntu.yml]: Is this a requirement of the new versions of TF/Keras?

File: openfl/federated/task/runner_keras.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ def rebuild_model(self, round_num, input_tensor_dict, validation=False):
         """
         if self.opt_treatment == "RESET":
             self.reset_opt_vars()
-            self.set_tensor_dict(input_tensor_dict, with_opt_vars=False)
+            self.set_tensor_dict(input_tensor_dict, with_opt_vars=True)
```
Comment (teoparvanov) [inline openfl/federated/task/runner_keras.py]: Intuitively, if the opt treatment is "RESET", shouldn't we have `with_opt_vars=False` ?

File: openfl/federated/task/runner_keras.py [hunk 3]
```diff
@@ -281,7 +280,12 @@ def _get_weights_names(obj):
         Returns:
             weight_names (list): The weight name list.
         """
-        weight_names = [weight.name for weight in obj.weights]
+        if isinstance(obj, ke.optimizers.Optimizer):
+            weight_names = [weight.name for weight in obj.variables]
+        else:
+            weight_names = [
+                layer.name + "/" + weight.name for layer in obj.layers for weight in layer.weights
+            ]
```
Comment (teoparvanov) [inline openfl/federated/task/runner_keras.py:287]: Which other optimizers would trigger the `else` block?

File: openfl/federated/task/runner_keras.py [hunk 4]
```diff
@@ -298,10 +302,22 @@ def _get_weights_dict(obj, suffix=""):
...
+        if isinstance(obj, ke.optimizers.Optimizer):
+            weight_names = [weight.name for weight in obj.variables]
+            weights_dict = {
+                weight_names[i] + suffix: weight.numpy()
+                for i, weight in enumerate(copy.deepcopy(obj.variables))
+            }
+        else:
+            weight_names = [
+                layer.name + "/" + weight.name for layer in obj.layers for weight in layer.weights
+            ]
```
Comment (teoparvanov) [inline openfl/federated/task/runner_keras.py]: This code is repeated several times - could you extract it in a separate function?

File: openfl-workspace/keras_nlp/requirements.txt
Comment (teoparvanov) [inline openfl-workspace/keras_nlp/requirements.txt:1]: I don't think the `keras_nlp` is covered by the CI. Have you tested it locally after upgrading to Keras 3?

**Review by teoparvanov — COMMENTED** [review]
LGTM overall. A couple of questions came up during my read-through:

File: .github/workflows/ubuntu.yml (hunk 1 above)
Comment (tanwarsh) [reply .github/workflows/ubuntu.yml]: yes, tensorflow and keras are not compatible with python < 3.9

File: openfl/federated/task/runner_keras.py (hunk 2 above)
Comment (tanwarsh) [reply openfl/federated/task/runner_keras.py]: This is done as when optmizers weights are reset. Leraning rate is zero and model accuracy remiains around ~0.125.

File: openfl/federated/task/runner_keras.py (hunk 3 above)
Comment (tanwarsh) [reply openfl/federated/task/runner_keras.py:287]: else is for model weights and if is for Optimizer weights

File: openfl-workspace/keras_nlp/requirements.txt
Comment (tanwarsh) [reply openfl-workspace/keras_nlp/requirements.txt:1]: Yes, I have tested Keras nlp and Kears cnn mnist both locally.

File: openfl/federated/task/runner_keras.py (hunk 4 above)
Comment (tanwarsh) [reply openfl/federated/task/runner_keras.py]: removed duplicate code.

File: .github/workflows/pki.yml [hunk 5]
```diff
@@ -22,10 +22,10 @@ jobs:
 
     steps:
     - uses: actions/checkout@v3
-    - name: Set up Python 3.8
+    - name: Set up Python 3.9
```
Comment (payalcha) [inline .github/workflows/pki.yml]: Please change the task_runner_e2e.yaml workflow as well. It runs nightly so not failed for the PR.

File: .github/workflows/double_ws_export.yml [hunk 6]
```diff
@@ -22,10 +22,10 @@ jobs:
 
     steps:
     - uses: actions/checkout@v3
-    - name: Set up Python 3.8
+    - name: Set up Python 3.9
```
Comment (payalcha) [inline .github/workflows/double_ws_export.yml]: If 3.8 not supported anymore, change the README documents as well for the same.

File: openfl-workspace/keras_cnn_mnist/src/taskrunner.py [hunk 7]
```diff
@@ -3,11 +3,10 @@
 
 """You may copy this file as the starting point of your own model."""
 
-import tensorflow.keras as ke
-from tensorflow.keras import Sequential
-from tensorflow.keras.layers import Conv2D
-from tensorflow.keras.layers import Dense
-from tensorflow.keras.layers import Flatten
+from keras.models import Sequential
```
Comment (payalcha) [inline openfl-workspace/keras_cnn_mnist/src/taskrunner.py:6]: Tensorflow Keras is still evident and we see many examples in the ML community for the same.
Shouldn't we add separate workspace for Keras 3.0 for now and remove this eventually. 

Or we are saying that we are not going to support tensorflow keras anymore.

File: openfl/federated/task/runner_keras.py [hunk 8]
```diff
@@ -17,8 +18,8 @@
 
 with catch_warnings():
     simplefilter(action="ignore")
+    import keras as ke
     import tensorflow as tf
```
Comment (payalcha) [inline openfl/federated/task/runner_keras.py]: Ideally tensorflow must not be used with Keras as Keras implementation must be free of any backend while running.
Please find the alternative for the same.

File: .github/workflows/pki.yml (hunk 5 above)
Comment (tanwarsh) [reply .github/workflows/pki.yml]: done

File: openfl/federated/task/runner_keras.py [hunk 9]
```diff
@@ -348,46 +361,23 @@ def set_tensor_dict(self, tensor_dict, with_opt_vars):
...
+            opt_weight_names = self._get_weights_names(self.model.optimizer)
             opt_weights_dict = {name: tensor_dict[name] for name in opt_weight_names}
             self._set_weights_dict(self.model, model_weights_dict)
             self._set_weights_dict(self.model.optimizer, opt_weights_dict)
 
     def reset_opt_vars(self):
         """Resets the optimizer variables."""
-        for var in self.model.optimizer.variables():
+        for var in self.model.optimizer.variables:
             var.assign(tf.zeros_like(var))
```
Comment (kminhta) [inline openfl/federated/task/runner_keras.py]: Re: learning rate resetting to 0
This line seems to be the cause of that issue. It's a legacy line so I am not sure the context behind it, but for some reason this line will reset every variable of the optimizer (including learning rate) to 0, which is why the model is not learning when `self.opt_treatment == "RESET"`. The torch task runner actually seems to just pass through this method entirely ([see](https://github.com/securefederatedai/openfl/blob/develop/openfl/federated/task/runner_pt.py#L453))

FYI, we should also get rid of any `tf` calls or imports to remove any dependency on tensorflow in the keras task runner

File: openfl/federated/task/runner_keras.py (hunk 2 above)
Comment (kminhta) [reply openfl/federated/task/runner_keras.py]: Once `reset_opt_vars()` is fixed, let's set this back to `with_opt_vars=False` since the intention is to not rebuild the model with the global optimizer (and "reset" it back to the local optimizer)

File: openfl/federated/task/runner_keras.py (hunk 8 above)
Comment (kminhta) [inline openfl/federated/task/runner_keras.py]: we should get rid of this import. I think the only place where `tf` is called is in `reset_opt_vars()`

File: openfl-workspace/keras_cnn_mnist/src/taskrunner.py (hunk 7 above)
Comment (kminhta) [reply openfl-workspace/keras_cnn_mnist/src/taskrunner.py:6]: You're right that we should not remove support for tf-keras. We will eventually update the existing tf workspaces to tf-keras and deprecate native tf support since keras has been adopted as the "official" high-level API for tf. 

The goal in this PR is to have an entirely native keras task runner and accompanying workspace since we will eventually want to support jax-keras and torch-keras as well.

**Comment by kminhta** [comment]
> @teoparvanov @kta-intel @psfoley, I was facing an issue with optimizer weights due to resetting the weights before saving them to the tensor database and using them to rebuild the model, which resulted in no improvement in accuracy. I have skipped the rebuild (and hence resetting the model optimizer weights) for the first round. This can be further discussed, and we can proceed with these changes once all the pipelines are working and the workspace changes are also completed.

Despite the name, I think `rebuild_model()` is actually a necessary step in the first round since the flow begins with the aggregator sending out a global model, so the collaborators do need to rebuild their initially compiled models with the global model's weights. We should be able to bring it back once the comments around `reset_opt_vars()` are addressed

File: openfl/federated/task/runner_keras.py [hunk 10]
```diff
@@ -17,8 +18,8 @@
 
 with catch_warnings():
     simplefilter(action="ignore")
+    import keras as ke
```
Comment (psfoley) [inline openfl/federated/task/runner_keras.py]: I realize this is carried over from prior examples, but `ke` is not a standard abbreviation for `keras`: https://keras.io/examples/vision/image_classification_with_vision_transformer/

While all of the keras code is being updated, I would recommend changing this to the following to make the code slightly easier to read for developers:

```suggestion
    import keras
```

File: openfl/federated/task/runner_keras.py [hunk 11]
```diff
@@ -98,7 +98,8 @@ def train(
             raise KeyError("metrics must be defined")
 
         # rebuild model with updated weights
-        self.rebuild_model(round_num, input_tensor_dict)
+        if round_num > 0:
```
Comment (psfoley) [inline openfl/federated/task/runner_keras.py]: I believe this will result in the aggregator's initial model not being used for the first round. This will be problematic if a pretrained model is used as the basis for federated fine-tuning.

File: openfl/federated/task/runner_keras.py [hunk 12]
```diff
@@ -220,23 +222,20 @@ def validate(self, col_name, round_num, input_tensor_dict, **kwargs):
         else:
             batch_size = 1
 
-        self.rebuild_model(round_num, input_tensor_dict, validation=True)
+        if round_num > 0:
```
Comment (psfoley) [inline openfl/federated/task/runner_keras.py]: This should also be reverted.

**Review by psfoley — DISMISSED** [review]

File: .github/workflows/double_ws_export.yml (hunk 6 above)
Comment (psfoley) [reply .github/workflows/double_ws_export.yml]: Keras 3.0 drops python 3.8 support, but other frameworks (i.e. PyTorch) still support it. If we decide to remove support for Python 3.8, it should be its own PR.

Comment (tanwarsh) [reply .github/workflows/double_ws_export.yml]: raised PR https://github.com/securefederatedai/openfl/pull/1189 for version change

File: openfl/federated/task/runner_keras.py (hunk 12 above)
Comment (tanwarsh) [reply openfl/federated/task/runner_keras.py]: reverted.

File: openfl/federated/task/runner_keras.py (hunk 11 above)
Comment (tanwarsh) [reply openfl/federated/task/runner_keras.py]: reverted.

File: openfl/federated/task/runner_keras.py (hunk 10 above)
Comment (tanwarsh) [reply openfl/federated/task/runner_keras.py]: done

File: openfl/federated/task/runner_keras.py (hunk 8 above)
Comment (tanwarsh) [reply openfl/federated/task/runner_keras.py]: done

File: openfl/federated/task/runner_keras.py (hunk 2 above)
Comment (tanwarsh) [reply openfl/federated/task/runner_keras.py]: done

File: openfl/federated/task/runner_keras.py (hunk 9 above)
Comment (tanwarsh) [reply openfl/federated/task/runner_keras.py]: this code is removed.

File: openfl-workspace/keras_cnn_mnist/src/taskrunner.py (hunk 7 above)
Comment (tanwarsh) [reply openfl-workspace/keras_cnn_mnist/src/taskrunner.py:6]: workspaces for older keras are removed/updated in this PR and  support for tf - keras will be marked deplrecated in https://github.com/securefederatedai/openfl/pull/1204 and eventually removed in future.

File: openfl/federated/task/runner_keras.py (hunk 8 above)
Comment (tanwarsh) [reply openfl/federated/task/runner_keras.py]: tensorflow is removed.

File: openfl-workspace/keras_cnn_mnist/plan/plan.yaml [hunk 13]
```diff
@@ -1,13 +1,13 @@
...
 aggregator :
   defaults : plan/defaults/aggregator.yaml
   template : openfl.component.Aggregator
   settings :
-    init_state_path : save/keras_cnn_mnist_init.pbuf
-    best_state_path : save/keras_cnn_mnist_best.pbuf
-    last_state_path : save/keras_cnn_mnist_last.pbuf
+    init_state_path : save/init.pbuf
+    best_state_path : save/best.pbuf
+    last_state_path : save/last.pbuf
```
Comment (MasterSkepticista) [inline openfl-workspace/keras_cnn_mnist/plan/plan.yaml:10]: Overall: Good that you generalized paths and module names throughout the PR.

File: openfl-workspace/keras_cnn_mnist/src/taskrunner.py
Comment (MasterSkepticista) [inline openfl-workspace/keras_cnn_mnist/src/taskrunner.py:1]: Since this is Keras3, what is the default backend? 

My suggestion would be to set the backend as an env variable explicitly. That informs the user how to switch the backends. It would also be good to test all 3 backends with the same code and ensure it passes (because OpenFL is agnostic to backend)

File: openfl-workspace/keras_nlp/src/dataloader.py [hunk 14]
```diff
@@ -68,7 +66,7 @@ def get_train_loader(self, batch_size: Optional[int] = None) -> Iterator[List[np
         return self._get_batch_generator(X1=self.X_train[0], X2=self.X_train[1],
                                          y=self.y_train, batch_size=batch_size)
 
-    def get_valid_loader(self, batch_size: Optional[int] = None) -> Iterator[List[np.ndarray]]:
+    def get_valid_loader(self, batch_size: Optional[int] = None):
```
Comment (MasterSkepticista) [inline openfl-workspace/keras_nlp/src/dataloader.py]: Returns a callable instead?

File: openfl/federated/task/runner_keras_tf.py [hunk 15]
```diff
@@ -0,0 +1,480 @@
+# Copyright 2020-2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+
+"""
+Base classes for developing a ke.Model() Federated Learning model.
```
Comment (theakshaypant) [inline openfl/federated/task/runner_keras_tf.py]: Nit: Remove the `ke` acronym.

File: openfl-workspace/keras_nlp/src/taskrunner.py
Comment (MasterSkepticista) [inline openfl-workspace/keras_nlp/src/taskrunner.py:1]: A reader cannot tell what this example is about (model/use case). Suggest adding a header (also making the license header consistent with other files) that tells what architecture, and what use case this example is for.

**Review by theakshaypant — COMMENTED** [review]
Is openfl setup not affected by these changes?

File: openfl/federated/__init__.py [hunk 16]
```diff
@@ -14,7 +14,11 @@
     from openfl.federated.data import FederatedDataSet  # NOQA
     from openfl.federated.data import KerasDataLoader, TensorFlowDataLoader
     from openfl.federated.task import FederatedModel  # NOQA
-    from openfl.federated.task import KerasTaskRunner, TensorFlowTaskRunner
+    from openfl.federated.task import (
+        KerasTaskRunner,
+        TensorFlowTaskRunner,
+        KerasTensorFlowTaskRunner,
```
Comment (MasterSkepticista) [inline openfl/federated/__init__.py]: ~Rename all instances of `KerasTensorFlowTaskRunner` to `TensorFlowKerasTaskRunner`. It may feel like a nitpick, but the legacy API namespace is actually `tf.keras`.~

~Current naming may confuse the user to interpret `KerasTensorFlowTaskRunner` as "Tensorflow backend of Keras3"~

Edit: OK, I am confused now. What exactly are we supporting? I thought this was about updating to `KerasTaskRunner` (i.e. the multi-backend Keras3) and `TensorFlowKerasTaskRunner` (i.e. the legacy tf 2.x namespace). Please clarify?

**Comment by MasterSkepticista** [comment]
I think this is a good place to refactor the directory structure:

```shell
openfl-workspace (in future, "examples")
 |_ keras
 |  |_ mnist
 |  |_ lstm
 |_ (other backends in future PRs)
```

Thoughts?

**Comment by teoparvanov** [comment]
> I think this is a good place to refactor the directory structure:
> 
> ```shell
> openfl-workspace (in future, "examples")
>  |_ keras
>  |  |_ mnist
>  |  |_ lstm
>  |_ (other backends in future PRs)
> ```
> 
> Thoughts?

@MasterSkepticista , would it be acceptable to do this in a follow-up PR?

File: openfl/federated/task/runner_keras_tf.py (hunk 15 above)
Comment (tanwarsh) [reply openfl/federated/task/runner_keras_tf.py]: Removed.

File: openfl-workspace/keras_cnn_mnist/src/taskrunner.py
Comment (tanwarsh) [reply openfl-workspace/keras_cnn_mnist/src/taskrunner.py:1]: By default it is tensorflow. This is already planned. There will be followup PR which will restructure code to support all backends (tensorflow, JAX and Torch)

File: openfl/federated/__init__.py (hunk 16 above)
Comment (tanwarsh) [reply openfl/federated/__init__.py]: - KerasTaskRunner: Keras 2 and TensorFlow 2.13
- TensorFlowTaskRunner: Keras 2 and TensorFlow 2.13
- KerasTensorFlowTaskRunner: Updated to Keras 3 and TensorFlow 2.18

KerasTaskRunner is used by 5 workspaces prior to this PR:

- Keras_cnn_mnist
- Keras_cnn_with_compression
- Keras_nlp
- tf_3dunet_brats
- tf_cnn_histology

TensorFlowTaskRunner is used by tf_2dunet.

After this PR and https://github.com/securefederatedai/openfl/pull/1204 we will have all 6 workspaces updated with the latest Keras and TensorFlow versions. https://github.com/securefederatedai/openfl/pull/1204 will also move older TensorFlow workspace code to the deprecated folder, which will be removed in the future. Therefore, KerasTaskRunner and TensorFlowTaskRunner will be removed later.

File: openfl-workspace/keras_nlp/src/dataloader.py (hunk 14 above)
Comment (tanwarsh) [reply openfl-workspace/keras_nlp/src/dataloader.py]: fixed.

File: openfl/federated/task/runner_keras_tf.py [hunk 17]
```diff
@@ -0,0 +1,480 @@
...
+from openfl.utilities import Metric, TensorKey, change_tags
+from openfl.utilities.split import split_tensor_dict_for_holdouts
+
+with catch_warnings():
+    simplefilter(action="ignore")
+    import keras
+
+
+class KerasTensorFlowTaskRunner(TaskRunner):
+    """The base model for Keras models in the federation.
```
Comment (teoparvanov) [inline openfl/federated/task/runner_keras_tf.py]: nit - base task runner

**Review by teoparvanov — APPROVED** [review]
Great work, thanks @tanwarsh ! This PR brings OpenFL back into the TensorFlow game 😉

File: openfl/federated/__init__.py (hunk 16 above)
Comment (kminhta) [reply openfl/federated/__init__.py]: I might've missed some discussion. What is the rationale for keeping `KerasTaskRunner` and introducing a new `KerasTensorFlowTaskRunner` instead of just updating `KerasTaskRunner` as a multi-backend Keras3 (as @MasterSkepticista mentioned)?

If we have large enough justification to keep Keras 2 support in this release, we should just add another class to `runner_keras.py`, imo. After this and that other PR you mentioned, it would look like:
- `runner_keras.KerasTaskRunner` -> supports Keras 3 and TensorFlow 2.18
- `runner_keras.Keras2TaskRunner` -> supports Keras 2 and TensorFlow 2.13
- remove/deprecate `TensorFlowTaskRunner`

`runner_keras.KerasTaskRunner` should support anything that Keras supports (i.e. TF, keras-JAX, keras-torch, etc.), so we won't need a separate `KerasTensorFlowTaskRunner`

Comment (MasterSkepticista) [reply openfl/federated/__init__.py]: > * KerasTaskRunner: Keras 2 and TensorFlow 2.13
> * TensorFlowTaskRunner: Keras 2 and TensorFlow 2.13
> * KerasTensorFlowTaskRunner: Updated to Keras 3 and TensorFlow 2.18
> 
> KerasTaskRunner is used by 5 workspaces prior to this PR:
> 
> * Keras_cnn_mnist
> * Keras_cnn_with_compression
> * Keras_nlp
> * tf_3dunet_brats
> * tf_cnn_histology
> * TensorFlowTaskRunner is used by tf_2dunet.
> 
> After this PR and #1204 we will have all 6 workspaces updated with the latest Keras and TensorFlow versions. #1204 will also move older TensorFlow workspace code to the deprecated folder, which will be removed in the future. Therefore, KerasTaskRunner and TensorFlowTaskRunner will be removed later.

I remain confused with this. Here is my proposal:

* `KerasTaskRunner`: A new Keras 3 task runner which is agnostic of any backend that "keras" provides support for, i.e., any keras 3 taskrunner workspace should work with torch/tf/jax/numpy backends.
* `TensorFlowTaskRunner`: The `tf.keras` TaskRunner for `tensorflow<=2.16,>=2.0`. This can remain in maintenance mode for about a year, because community no longer writes entire NNs in pure TensorFlow anymore. They may write using `tf.keras`, but Google will likely deprecate this fully next year.
* `PyTorchTaskRunner`: as-is.

Note: This may mean removing support (+code/examples) for `tensorflow==1.x`, which is what we refer by `TensorFlowTaskRunner` today. `tf1.x` API is so old and practically unusable, there is no harm removing it without looking back.

Cc: @kta-intel, https://github.com/securefederatedai/openfl/pull/985#issuecomment-2163817246 https://github.com/securefederatedai/openfl/pull/985#issuecomment-2163323526

Comment (MasterSkepticista) [reply openfl/federated/__init__.py]: > I might've missed some discussion. What is the rationale for keeping `KerasTaskRunner` and introducing a new `KerasTensorFlowTaskRunner` instead of just updating `KerasTaskRunner` as a multi-backend Keras3 (as @MasterSkepticista mentioned)?
> 
> If we have large enough justification to keep Keras 2 support in this release, we should just add another class to `runner_keras.py`, imo. After this and that other PR you mentioned, it would look like:
> 
> * `runner_keras.KerasTaskRunner` -> supports Keras 3 and TensorFlow 2.18
> * `runner_keras.Keras2TaskRunner` -> supports Keras 2 and TensorFlow 2.13
> * remove/deprecate `TensorFlowTaskRunner`
> 
> `runner_keras.KerasTaskRunner` should support anything that Keras supports (i.e. TF, keras-JAX, keras-torch, etc.), so we won't need a separate `KerasTensorFlowTaskRunner`

I agree, Kevin. One tiny detail though: Keras 2 isn't commonly known, except for us since we closely track tf2 and keras3 naming. Users would only know `tf.keras` and `keras>=3`. So my suggestion (subject to having a consensus) would be to deprecate `TensorFlowTaskRunner` as you mentioned, since it refers to `tf1.x`, but retain the same name for Keras2 (equivalently `tf.keras`).

TL;DR: We deprecate all `tf1.x` code in OpenFL, but retain the class name `TensorFlowTaskRunner` for all `tf2.x` code. 

WDYT?

Comment (kminhta) [reply openfl/federated/__init__.py]: > I agree, Kevin. One tiny detail though: Keras 2 isn't commonly known, except for us since we closely track tf2 and keras3 naming. Users would only know `tf.keras` and `keras>=3`. So my suggestion (subject to having a consensus) would be to deprecate `TensorFlowTaskRunner` as you mentioned, since it refers to `tf1.x`, but retain the same name for Keras2 (equivalently `tf.keras`).
> 
> TL;DR: We deprecate all `tf1.x` code in OpenFL, but retain the class name `TensorFlowTaskRunner` for all `tf2.x` code.
> 
> WDYT?

I like this proposal, it's a good idea. +1 from me!
Using `TensorFlowTaskRunner` as our vehicle for Keras2 support via `tf.keras` makes a lot more sense

**Comment by tanwarsh** [comment]
> Is openfl setup not affected by these changes?

No this will not affect openfl setup.

**Comment by tanwarsh** [comment]
@teoparvanov @MasterSkepticista @kta-intel 

As discussed I have migrated workspaces mentioned below to latest tensorflow and keras 3 versions and rest of the workspaces are removed.

1. keras_cnn_mnist
2. keras_nlp
3. keras_2dunet (renamed from tf_2dunet)

File: openfl/federated/__init__.py (hunk 16 above)
Comment (tanwarsh) [reply openfl/federated/__init__.py]: TensorFlowTaskRunner  and KerasTensorFlowTaskRunner code removed. KerasTaskRunner is update to support latest versions.

**Review by rahulga1 — APPROVED** [review]

**Review by teoparvanov — APPROVED** [review]
LGTM!
